rule/unconditional-recursion.go

```diff
@@ -0,0 +1,187 @@
+
+// Apply applies the rule to given file.
+func (r *UnconditionalRecursionRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
+	var failures []lint.Failure
+
+	onFailure := func(failure lint.Failure) {
+		failures = append(failures, failure)
+	}
+
+	// err := file.Pkg.TypeCheck()
+	// if err != nil {
+	// 	panic(fmt.Sprintf("Error while type-checking file %s: %v", file.Name, err))
+	// }
+	w := lintUnconditionalRecursionRule{onFailure: onFailure}
+	ast.Walk(w, file.AST)
+	return failures
+}
+
+// Name returns the rule name.
+func (r *UnconditionalRecursionRule) Name() string {
+	return "unconditional-recursion"
+}
+
+type funcDesc struct {
+	reciverId *ast.Ident
+	id        *ast.Ident
+}
+
+func (fd *funcDesc) equal(other *funcDesc) bool {
+	receiversAreEqual := (fd.reciverId == nil && other.reciverId == nil) || fd.reciverId != nil && other.reciverId != nil && fd.reciverId.Name == other.reciverId.Name
+	idsAreEqual := (fd.id == nil && other.id == nil) || fd.id.Name == other.id.Name
+
+	return receiversAreEqual && idsAreEqual
+}
+
+type funcStatus struct {
+	funcDesc            *funcDesc
+	seenConditionalExit bool
+}
+
+type lintUnconditionalRecursionRule struct {
+	onFailure   func(lint.Failure)
+	currentFunc *funcStatus
+}
+
+// Visit will traverse the file AST.
+// The rule is based in the following algorithm: inside each function body we search for calls to the function itself.
+// We do not search inside conditional control structures (if, for, switch, ...) because any recursive call inside them is conditioned
+// We do search inside conditional control structures are statements that will take the control out of the function (return, exit, panic)
+// If we find conditional control exits, it means the function is NOT unconditionally-recursive
+// If we find a recursive call before finding any conditional exit, a failure is generated
+// In resume: if we found a recursive call control-dependant from the entry point of the function then we raise a failure.
+func (w lintUnconditionalRecursionRule) Visit(node ast.Node) ast.Visitor {
+	switch n := node.(type) {
+	case *ast.FuncDecl:
+		var rec *ast.Ident
+		switch {
+		case n.Recv == nil || n.Recv.NumFields() < 1 || len(n.Recv.List[0].Names) < 1:
+			rec = nil
+		default:
+			rec = n.Recv.List[0].Names[0]
+		}
+
+		w.currentFunc = &funcStatus{&funcDesc{rec, n.Name}, false}
+	case *ast.CallExpr:
+		var funcId *ast.Ident
+		var selector *ast.Ident
+		switch c := n.Fun.(type) {
+		case *ast.Ident:
+			selector = nil
+			funcId = c
+		case *ast.SelectorExpr:
+			var ok bool
+			selector, ok = c.X.(*ast.Ident)
+			if !ok { // a.b....Foo()
+				return nil
+			}
+			funcId = c.Sel
+		default:
+			return w
+		}
+
+		if w.currentFunc != nil && // not in a func body
+			!w.currentFunc.seenConditionalExit && // there is a conditional exit in the function
+			w.currentFunc.funcDesc.equal(&funcDesc{selector, funcId}) {
+			w.onFailure(lint.Failure{
+				Category:   "logic",
+				Confidence: 1,
```

Comment:
Maybe we can decrease the confidence a bit? A construct with a for loop when we're draining a channel may still throw?

Reply:
The rule is all about control dependencies.
`for-range`'s bodies are analysed looking for stmt taking the control out of the function (`return`, `os.Exit`, ...) to check if subsequent statements depend upon the loop
A channel-draining loop is not different from other loop.

About the confidence: I do not see cases where we can generate a false positive (do you have examples in mind). Contrary we can pass silently recursive calls, for example:
```go
func foo() {
  for false { // or other arbitrary complex condition evaluating to false
    return
  }
  foo()
``` 
will not trigger a failure.